src/repl.ts

```diff
@@ -612,6 +612,37 @@ async function selectJuliaBlock() {
     }
 }
 
+async function executeJuliaCellInRepl() {
+    telemetry.traceEvent('command-executejuliacellinrepl');
+
+    let ed = vscode.window.activeTextEditor;
+    let doc = ed.document;
+    let rx = new RegExp("^(##|#---|#%%|# %%)"); // atom-julia-client defaults
+    let curr = ed.selection.active.line;
+    var start = curr;
+    while (start >= 0) {
+        if (rx.test(doc.lineAt(start).text)) {
+            break;
+        } else {
+            start -= 1;
+        }
+    }
+    start += 1;
+    var end = start;
+    while (end < doc.lineCount) {
+        if (rx.test(doc.lineAt(end).text)) {
+            break;
+        } else {
+            end += 1;
+        }
+    }
+    end -= 1;
+    let startpos = new vscode.Position(start, 0);
+    let endpos = new vscode.Position(end, doc.lineAt(end).text.length);
+    let code = doc.getText(new vscode.Range(startpos, endpos));
+    executeCode(code, false)
```

Comment:
I think we should use the `sendMessage('repl/runcode', msg_body)` pattern instead, that makes for a much more stable code execution story. It would probably make sense to create a new function a la `executeCode2` (name really doesn't matter much at this point) that wraps the logic to execute code that way. Right now there are two places in the code that use this ([here](https://github.com/julia-vscode/julia-vscode/blob/9a2a2fbd839b79ce67de02216fe184ba0208b72c/src/repl.ts#L588) and [here](https://github.com/julia-vscode/julia-vscode/blob/9a2a2fbd839b79ce67de02216fe184ba0208b72c/src/repl.ts#L637)), and they could then also use such a `executeCode2` function. All that function would do is essentially wrap [this](https://github.com/julia-vscode/julia-vscode/blob/9a2a2fbd839b79ce67de02216fe184ba0208b72c/src/repl.ts#L633) stuff.

The benefit of that approach is that the code executed that way has line numbers associated with it, and things like `@__DIR__` work.

Reply:
I pushed a version of this in my latest commit. This was the first thing I tried, but I wasn't happy without having a visible sign that the cell was executed. It would be nice to have something visible happen at the REPL.

Reply:
Yeah, I was unsure whether to echo code back or not. [This](https://github.com/julia-vscode/julia-vscode/blob/9a2a2fbd839b79ce67de02216fe184ba0208b72c/scripts/terminalserver/terminalserver.jl#L149) line does echo code back if it is not disabled.

I'm just a bit worried that it gets unwieldy when one executes a whole file or a lot of code... But maybe that would be ok. Or maybe we need some other visual identification that code is running.

@pfitzseb, are you always echoing all code back in the REPL that is being executed in Juno?

Reply:
I'm pretty sure we're *never* echoing executed code in the REPL.

> Or maybe we need some other visual identification that code is running.

VSCode does support custom status bar elements, right? Would maybe be useful to have something like Juno's [progress meter](http://docs.junolab.org/latest/man/juno_frontend/#Progress-Meters-1). Also, the REPL prompt is hidden when code is executed, so that should also be a good indicator that *something*'s happening.

Reply:
Yeah, I agree, a UI in the VS Code chrome would be nice. One thing, though: how could we hook into that if folks just type code into the REPL and run it? There are no callback hooks for the REPL a la "about to run user code", and "finished running user code", right?

Reply:
True, that's a bit messy.
Juno has [this abomination](https://github.com/JunoLab/Atom.jl/blob/92ad44109f24b80aa49f233dade4ae9579817f20/src/repl.jl#L139). I'm starting to think it might make sense to prototype some repl enhancements in a `REPLPlus.jl` package that could be used by Juno and VSCode and provides a bunch of hooks (and maybe even solves the whole async printing business properly).

Reply:
Ok, I looked at it, but I don't get how this helps with the user typing something in the REPL and pressing Enter. Does that code somehow hook into that?

> I'm starting to think it might make sense to prototype some repl enhancements...

Yes, that would be great! Or maybe we could even add these two simple hooks into julia base for a future version? If both VS Code and Juno could make use of such hooks, they shouldn't be too controversial?

Reply:
> Ok, I looked at it, but I don't get how this helps with the user typing something in the REPL and pressing Enter. Does that code somehow hook into that?

Yes, we're replacing the REPL `respond` function with that [here](https://github.com/JunoLab/Atom.jl/blob/92ad44109f24b80aa49f233dade4ae9579817f20/src/repl.jl#L187-L193).

> Yes, that would be great! Or maybe we could even add these two simple hooks into julia base for a future version? If both VS Code and Juno could make use of such hooks, they shouldn't be too controversial?

Sure, but Base/stdlibs are moving pretty slowly, so prototyping in an external package probabaly makes sense.

Reply:
Ah, ok, now I get it!

Yes, totally agreed, prototyping somewhere else probably is much easier.